src/OrleansServiceBus/Providers/Streams/EventHub/CachePressureMonitors/AveragingCachePressureMonitor.cs

```diff
@@ -0,0 +1,93 @@
+    /// over all pressure contribution
+    /// </summary>
+    public class AveragingCachePressureMonitor : ICachePressureMonitor
+    {
+        /// <summary>
+        /// Default flow control threshold
+        /// </summary>
+        public static readonly double DefaultThreshold = 1 / 3;
+        private static readonly TimeSpan checkPeriod = TimeSpan.FromSeconds(2);
+        private readonly Logger logger;
+
+        private double accumulatedCachePressure;
+        private double cachePressureContributionCount;
+        private DateTime nextCheckedTime;
+        private bool isUnderPressure;
+        private double flowControlThreshold;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="logger"></param>
+        public AveragingCachePressureMonitor(Logger logger)
+            :this(DefaultThreshold, logger)
+        { }
+
+        /// <summary>
+        /// Contructor
+        /// </summary>
+        /// <param name="flowControlThreshold"></param>
+        /// <param name="logger"></param>
+        public AveragingCachePressureMonitor(double flowControlThreshold, Logger logger)
+        {
+            this.flowControlThreshold = flowControlThreshold;
+            this.logger = logger.GetSubLogger(this.GetType().Name);
+            nextCheckedTime = DateTime.MinValue;
+            isUnderPressure = false;
+        }
+
+        public void RecordCachePressureContribution(double cachePressureContribution)
+        {
+            // Weight unhealthy contributions thrice as much as healthy ones.
+            // This is a crude compensation for the fact that healthy consumers wil consume more often than unhealthy ones.
+            double weight = cachePressureContribution < flowControlThreshold ? 1.0 : 3.0;
+            accumulatedCachePressure += cachePressureContribution * weight;
+            cachePressureContributionCount += weight;
+        }
+
+        public bool IsUnderPressure(DateTime utcNow)
+        {
+            if (nextCheckedTime < utcNow)
+            {
+                CalculatePressure();
+                nextCheckedTime = utcNow + checkPeriod;
+            }
+            return isUnderPressure;
+        }
+
+        private void CalculatePressure()
+        {
+            // if we don't have any contributions, don't change status
+            if (cachePressureContributionCount < 0.5)
+            {
+                // after 5 checks with no contributions, check anyway
+                cachePressureContributionCount += 0.1;
+                return;
+            }
+
+            double pressure = accumulatedCachePressure / cachePressureContributionCount;
+            bool wasUnderPressure = isUnderPressure;
+            isUnderPressure = pressure > flowControlThreshold;
+            // If we changed state, log
+            if (isUnderPressure != wasUnderPressure)
+            {
+                logger.Info(isUnderPressure
```

Comment:
I think the info level version of this log message should be in the aggregate pressure monitor, and the logging in the specific monitors should be verbose.
The reason for this is that a specific monitor may become healthy, but reading will still be off until all monitors are healthy.  So having per monitor logging will make it very hard for users to decipher, from logs, if a queue should be reading or not.  These details may be useful in verbose logging, as they would indicate what monitor is stopping reading, but this level of detail is less important, imo, than the overall flow control state.